File: javascript/src/repositories/billing/payment.ts
import { z } from "zod";
import type { QueryFn } from "../types.js";

export const BillingPaymentRowSchema = z
  .object({
    provider: z.string().optional(),
    provider_payment_id: z.string().optional(),
    user_id: z.string().nullable().optional(),
    amount_minor: z.number().optional(),
    tax_minor: z.number().nullable().optional(),
    currency: z.string().optional(),
    purpose: z.string().nullable().optional(),
    metadata: z.record(z.string(), z.unknown()).nullable().optional(),
    created_at: z.unknown().optional(),
    updated_at: z.unknown().optional(),
    credits_per_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    credits_per_major_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
  })
  .passthrough();

export const ForRefundRowSchema = z
  .object({
    credits_per_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    credits_per_major_unit: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    tier: z.string().optional(),
    deposit_to: z.string().optional(),
  })
  .passthrough();

export type BillingPaymentRow = z.infer<typeof BillingPaymentRowSchema>;
export type ForRefundRow = z.infer<typeof ForRefundRowSchema>;

function unwrapJsonb(rows: unknown[]): Record<string, unknown> | null {
  if (rows.length !== 1) return null;
  const row = rows[0] as Record<string, unknown>;
  const keys = Object.keys(row);
  if (keys.length !== 1) return null;
  const v = row[keys[0]];
  if (v !== null && typeof v === "object" && !Array.isArray(v)) return v as Record<string, unknown>;
  return null;
}

export class BillingPaymentRepository {
  constructor(private query: QueryFn) {}

  async upsert(
    provider: string,
    providerPaymentId: string,
    providerInvoiceId: string | null,
    userId: string | null,
    amountMinor: number,
    taxMinor: number | null,
    currency: string,
    purpose: string | null,
    metadata: string | null,
  ): Promise<void> {
    await this.query(`SELECT public.upsert_billing_payment($1, $2, $3, $4, $5, $6, $7, $8, $9)`, [
      provider,
      providerPaymentId,
      providerInvoiceId,
      userId,
      amountMinor,
      taxMinor,
      currency,
      purpose,
      metadata,
    ]);
  }

  async getForRefund(provider: string, providerPaymentId: string): Promise<ForRefundRow | null> {
    const rows = await this.query("SELECT * FROM public.get_billing_payment_for_refund($1, $2)", [
      provider,
      providerPaymentId,
    ]);
    const data = unwrapJsonb(rows);
    return data ? ForRefundRowSchema.parse(data) : null;
  }

  async getDirect(provider: string, providerPaymentId: string): Promise<BillingPaymentRow | null> {
    const rows = await this.query(
      `SELECT provider, provider_payment_id, user_id, amount_minor,
              tax_minor, currency, purpose, metadata, created_at, updated_at
       FROM public.billing_payments
       WHERE provider = $1 AND provider_payment_id = $2`,
      [provider, providerPaymentId],
    );
    if (rows.length === 0) return null;
    return BillingPaymentRowSchema.parse(rows[0]);
  }
}

File: python/src/bursar/repositories/billing/payment.py
from __future__ import annotations

from typing import Any

from bursar.repositories._types import QueryFn
from bursar.repositories._utils import unwrap_jsonb, validate_non_empty
from bursar.repositories.schemas import BillingPaymentRow


class BillingPaymentRepository:
    """Repository for billing payment operations.

    All methods call Postgres via raw SQL queries through the query function.
    Returns None when the query returns no rows.
    """

    def __init__(self, execute: QueryFn) -> None:
        self._execute = execute

    def upsert(
        self,
        provider: str,
        provider_payment_id: str,
        provider_invoice_id: str | None,
        user_id: str | None,
        amount_minor: int,
        tax_minor: int | None,
        currency: str,
        purpose: str | None,
        metadata: str | None,
    ) -> None:
        """Insert or update a billing payment record.

        Args:
            provider: The billing provider identifier.
            provider_payment_id: The provider payment ID.
            provider_invoice_id: The associated provider invoice ID, or None.
            user_id: The user ID, or None.
            amount_minor: The payment amount in minor currency units.
            tax_minor: The tax amount in minor currency units, or None.
            currency: The ISO 4217 currency code.
            purpose: The payment purpose, or None.
            metadata: JSON metadata string, or None.
        """
        self._execute(
            "SELECT public.upsert_billing_payment(%s, %s, %s, %s, %s, %s, %s, %s, %s)",
            [
                provider,
                provider_payment_id,
                provider_invoice_id,
                user_id,
                amount_minor,
                tax_minor,
                currency,
                purpose,
                metadata,
            ],
        )

    def get_for_refund(self, provider: str, provider_payment_id: str) -> dict[str, Any] | None:
        """Get payment details for refund processing.

        Args:
            provider: The billing provider identifier.
            provider_payment_id: The provider payment ID.

        Returns:
            Payment details dict if found, None otherwise.
        """
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_payment_id, "provider_payment_id")
        return unwrap_jsonb(
            self._execute(
                "SELECT * FROM public.get_billing_payment_for_refund(%s, %s)",
                [provider, provider_payment_id],
            )
        )

    def get_direct(self, provider: str, provider_payment_id: str) -> BillingPaymentRow | None:
        """Get a billing payment directly from the payments table.

        Args:
            provider: The billing provider identifier.
            provider_payment_id: The provider payment ID.

        Returns:
            BillingPaymentRow if found, None otherwise.
        """
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_payment_id, "provider_payment_id")
        rows = self._execute(
            """SELECT provider, provider_payment_id, user_id, amount_minor,
                      tax_minor, currency, purpose, metadata, created_at, updated_at
               FROM public.billing_payments
               WHERE provider = %s AND provider_payment_id = %s""",
            [provider, provider_payment_id],
        )
        if not rows:
            return None
        return BillingPaymentRow.model_validate(rows[0]) if isinstance(rows[0], dict) else None
